### PR TITLE
fix(ngcc): handle missing sources and out of range mappings when flattening source-maps

### LINKED ISSUE #35709: Angular 9.0.3 ngcc fails to process ng2-dragula
# 🐞 bug report

### Affected Package
@angular/compiler-cli

### Is this a regression?
Yes it works properly with angular 9.0.2

### Description
ngcc build failed with ng2-dragula.

## 🔬 Minimal Reproduction
1. Clone https://github.com/log2-hwan/ngcc-ng2-dragula-bug-demo
2. `npm install`
3. `npm run build ngcc-ng2-dragula-bug-demo`
4. Build failed with error

## 🔥 Exception or Error
<pre><code>
0% compiling
Compiling @angular/core : es2015 as esm2015

Compiling @angular/common : es2015 as esm2015

Compiling @angular/platform-browser : es2015 as esm2015

Compiling @angular/platform-browser-dynamic : es2015 as esm2015

Compiling ng2-dragula : es2015 as esm2015

ERROR in Cannot read property 'push' of undefined
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.0.3
Node: 13.8.0
OS: darwin x64

Angular: 9.0.3
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.3
@angular-devkit/build-angular     0.900.3
@angular-devkit/build-optimizer   0.900.3
@angular-devkit/build-webpack     0.900.3
@angular-devkit/core              9.0.3
@angular-devkit/schematics        9.0.3
@ngtools/webpack                  9.0.3
@schematics/angular               9.0.3
@schematics/update                0.900.3
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Anything else relevant?**
I guess the error is caused by commit df816c9 which is related to sourcemap processing, so ng2-dragula's sourcemap may have some problems.

### PR DESCRIPTION
**Pull request by petebacondarwin**
This PR has two commits that fix issues with source-maps that are badly formatted.

---

If a package has a source-map but it does not provide the actual content of the sources, then the source-map flattening was crashing.

Now we ignore such mappings that have no source since we are not able to compute the merged mapping if there is no source file.

---

Previously when rendering flattened source-maps, it was assumed that no mapping would come from a line that is outside the lines of the actual source content. It turns out this is not a valid assumption.

Now the code that renders flattened source-maps will handle such mappings, with the additional benefit that the rendered source-map will only contain mapping lines up to the last mapping, rather than a mapping line for every content line.

---

Fixes #35709

### CODE REVIEW

**Review by gkalpak — APPROVED** [review]
LGTM :+1:

**Comment by petebacondarwin** [comment]
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

**Comment by petebacondarwin** [comment]
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

**Comment by petebacondarwin** [comment]
The `ci/circleci: components-repo-unit-tests` is actually green - but GH didn't respond to the webhook from CircleCI.

File: packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts [hunk 1]
```diff
@@ -290,11 +295,16 @@ export function parseMappings(
     const generatedLineMappings = rawMappings[generatedLine];
     for (const rawMapping of generatedLineMappings) {
       if (rawMapping.length >= 4) {
+        const originalSource = sources[rawMapping[1] !];
+        if (originalSource === null) {
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts]: Could it have been an out-of-bounds read in which case it would be `undefined`?

Comment (alxhub) [reply packages/compiler-cli/ngcc/src/sourcemaps/source_file.ts]: I don't think it could hurt here doing `==` instead of `===` to catch this case.

**Review by alxhub — APPROVED** [review]

**Comment by gkalpak** [comment]
BTW, it would be a good idea to run this against angular/ngcc-validation before merging to make sure there are no other regressions.

**Comment by petebacondarwin** [comment]
Addressed the feedback. Now checking it against ngcc-validation repo.

File: packages/compiler-cli/ngcc/src/rendering/source_maps.ts [hunk 2]
```diff
@@ -24,28 +25,39 @@ export interface SourceMapInfo {
...
+    } else {
+      const sourceMapComment = generateMapFileComment(`${basename(generatedPath)}.map`);
+      return [
+        {path: generatedPath, contents: `${generatedFile.contents}\n${sourceMapComment}`},
+        {path: generatedMapPath, contents: mergedMap.toJSON()}
+      ];
+    }
+  } catch (e) {
+    logger.error(
+        `Failed to flatten the generated file source-map with the original source-maps failed: ${e.toString()}`);
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/rendering/source_maps.ts]:
```suggestion
        `Failed to flatten the generated file source-map with the original source-maps: ${e.toString()}`);
```

Comment (petebacondarwin) [reply packages/compiler-cli/ngcc/src/rendering/source_maps.ts]: Doh!

**Review by JoostK — APPROVED** [review]

**Pull request closed without merging**